### PR TITLE
Fix RT-DETR optional losses

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

In the modeling file of RT-DETR there are the following losses defined 
 - labels_varifocal (used by default)
 - boxes (used by default)
 - labels_focal (not used, broken)
 - labels_cross_entorpy (not used, broken)
 - labels_binary_cross_entropy (not used, broken)
 
This PR fixes additional losses that might be defined for fine-tuning and enable their usage for fine-tuning with config option

```python
model = RTDetrForObjectDetection.from_pretrained(
    "PekingU/rtdetr_r18vd",
    losses=["labels_varifocal", "boxes"],  # default value
    
    # Changing RT-DETR labels loss:
    # losses=["labels_focal", "boxes"],
    # losses=["labels_binary_cross_entropy", "boxes"],
    # losses=["labels_cross_entropy", "boxes"],
    
    # Combining multiple losses
    # losses=["labels_focal", "labels_varifocal", "boxes"],
)
```

I made several finetuning runs to make sure the model converges fine for every loss function
Screenshot 2024-08-13 at 12 19 22

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/rt_detr/modeling_rt_detr.py [hunk 1]
```diff
@@ -1970,40 +1971,61 @@ class RTDetrLoss(nn.Module):
...
+                f"Unsupported losses: {not_matched_losses}, should be in {self.losses_weight_dict.keys()}"
+            )
+
+        self.losses = config.losses
+
+        # for multiclass classfication with cross-entropy loss
+        if "labels_cross_entropy" in self.losses:
+            class_weights = torch.ones(config.num_labels + 1)
+            class_weights[-1] = config.eos_coefficient
+            self.register_buffer("class_weights", class_weights)
```
Comment (SangbumChoi) [inline src/transformers/models/rt_detr/modeling_rt_detr.py:2021]: Does this name change causes backward compability?

**Review by SangbumChoi — COMMENTED** [review]
Thanks for this detail changes! Overall it looks good.
Only thing that I'm concerned about is naming rule. All the other transformers model with object detection is named as `loss_{something}`.
So when we log the comput_metrics we can know that prefix indicates the type of value.
I like that writing `bce` as `binary_cross_entropy`. However, current change returns the loss value as `bbox`, `giou`, `labels_binary_cross_entropy`, etc...

File: src/transformers/models/rt_detr/modeling_rt_detr.py (hunk 1 above)
Comment (SangbumChoi) [reply src/transformers/models/rt_detr/modeling_rt_detr.py:2021]: + I think can we set both `labels_focal` and `labels_cross_entropy`? I don't think so

File: src/transformers/models/rt_detr/configuration_rt_detr.py [hunk 2]
```diff
@@ -328,13 +343,20 @@ def __init__(
...
         super().__init__(is_encoder_decoder=is_encoder_decoder, **kwargs)
 
     @property
     def num_attention_heads(self) -> int:
         return self.encoder_attention_heads
 
+    @property
+    def is_multiclass(self) -> bool:
+        return "labels_cross_entropy" in self.losses
+
```
[the remark this answers is not in the document]
Comment (dwchoo) [reply src/transformers/models/rt_detr/configuration_rt_detr.py:359]: Thank you for your thoughtful proposal and the efforts to enhance the model's flexibility. However, I respectfully disagree with the addition of a "void" class to the RT-DETR model. I believe this addition may contradict the original intentions behind the RT-DETR architecture.

I would greatly appreciate if you could review the comment I've referenced, which provides more context on this matter:
https://github.com/huggingface/transformers/pull/32579#issuecomment-2287550048

This comment offers insights into why the "void" class might not align with the core design principles of RT-DETR. I believe maintaining the model's original design without a "void" class would better preserve its intended functionality and architectural integrity.

I'm open to further discussion on this topic and would be interested in exploring alternative approaches that could enhance the model's capabilities while staying true to its original design philosophy. Your thoughts on this perspective would be greatly valued.

File: src/transformers/models/rt_detr/modeling_rt_detr.py (hunk 1 above)
Comment (qubvel) [reply src/transformers/models/rt_detr/modeling_rt_detr.py:2021]: yes,
1) as far as I see the loss is not a part of the model and is created for each iteration
2) at the moment `class_weights` is not used in loss computation, it only used in fixed losses

Comment (qubvel) [reply src/transformers/models/rt_detr/modeling_rt_detr.py:2021]: > Can we set both labels_focal and labels_cross_entropy? I don't think so

Technically we can do it, but not sure how good the results would be 🙂
